### PR TITLE
Hangman Functionality

### PR DESCRIPTION
**Pull request by zchbndcc9**
# About

For this implementation, I utilized a pipeline approach in which functions acted on the state depending on the current state. For this implementation, I didn't have time to do tests. If I had more time, I would add error handling and more test cases. In the future, I will definitely be starting sooner. Thank you

Zach Banducci 
44671228

## Update (Oct 1 - 1:50 AM)

Reworked some organization and fixed some bugs

## Update (Oct 1 - 2:47 AM)

Fixed some critical bugs for both `make_move` and `mix.exs` in the human_player directory

### CODE REVIEW

File: hangman/lib/hangman/game/game.ex [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  def check_already_guessed(game, guess) do
+    game.used
+      |> contains_letter?(guess)
+      |> already_guessed(game)
+  end
+
+  def already_guessed(true, game = %Hangman.Game{}),  do: %{game | game_state: :already_guessed}
+  def already_guessed(false, game = %Hangman.Game{}), do: %{game | game_state: :continue}
+
+  def check_guess(game = %Hangman.Game{game_state: :already_guessed}, _), do: game
```
Comment (pragdave) [inline hangman/lib/hangman/game/game.ex:48]: Yes! Not many people did the check for already guessed like this. It lets you use a pipeline, which is great.

### +3

File: hangman/lib/hangman/game/game.ex [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  def check_guess(game = %Hangman.Game{}, guess) do
+    game = %{game | last_guess: guess, used: [guess | game.used]}
+
+    game.word_to_guess
+      |> contains_letter?(guess)
+      |> letter_in_word(game, guess)
+  end
+
```
Comment (pragdave) [inline hangman/lib/hangman/game/game.ex:56]: You could just use `map` here :)

File: hangman/lib/hangman/game/game.ex [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  end
+  ## [ Helpers ]
+
+  def new_word() do
+    Dictionary.random_word()
+      |> String.split("", trim: true)
+  end
+
+  # Checks a list to determine whether or not it contains a letter
+  def contains_letter?(board, guess) do
```
Comment (pragdave) [inline hangman/lib/hangman/game/game.ex:85]: Or just `guess in board`

**Comment by pragdave** [comment]
I graded the application as it stood at midnight on the due date (ref d279e98537e440f016e65ecfedbea58080fae14a).

It had no tests, so I'm deducting 5. It also failed to pass my tests:

- it returned `'_'` for unknown letters, and not `"_"`
- `make_move` didn't return a tuple

I'm afraid that cost 20.

This is a shame: your code looked good. You just didn't give yourself enough time.

**Comment by zchbndcc9** [comment]
I totally understand that. I caught the tuple issue a little too late, which was fixed in my most recent PR. As for the `'_'` issue, I never would have caught that so I guess I should test more often. Thank you again

**Pull request closed without merging**